Source/SwiftLintBuiltInRules/Rules/Idiomatic/FinalClassRule.swift

```diff
@@ -0,0 +1,64 @@
+import SwiftSyntax
+
+@SwiftSyntaxRule
+struct FinalClassRule: Rule {
+    var configuration = SeverityConfiguration<Self>(.warning)
+
+    static let description = RuleDescription(
+        identifier: "final_class",
+        name: "Final Class",
+        description: "Prefer `static` over `final class`",
+        kind: .idiomatic,
+        nonTriggeringExamples: [
+            Example("""
+            class C {
+                static func f() {}
+            }
+            """),
+            Example("""
+            class C {
+                class func f() {}
+            }
+            """),
+            Example("""
+            final class C {}
+            """),
+        ],
+        triggeringExamples: [
+            Example("""
+            class C {
+                final class func f() {}
+            }
+            """),
+            Example("""
+            final class C {
+                class func f() {}
+            }
+            """)
+        ]
+    )
+}
+
+private extension FinalClassRule {
+    final class Visitor: ViolationsSyntaxVisitor<ConfigurationType> {
+        override func visit(_ node: FunctionDeclSyntax) -> SyntaxVisitorContinueKind {
+            if node.modifiers.contains(keyword: .final),
+               node.modifiers.contains(keyword: .class) {
+                violations.append(node.positionAfterSkippingLeadingTrivia)
+            }
+
+            return .visitChildren
+        }
+
+        override func visit(_ node: ClassDeclSyntax) -> SyntaxVisitorContinueKind {
+            if node.modifiers.contains(keyword: .final) {
+                node.memberBlock.members
```

Comment:
Better to maintain a stack of booleans that can be looked up to check whether the current scope is a final class. There are some examples of rules using a stack. Look for the `Stack` type and its references in rules.

Reply:
Ok cool, I have loosely followed an example I found in NestingRule.

A question about this approach though: to me it seems duplicative to build up a new stack (however minimal) to track information that can be derived directly from the existing syntax tree. The duplication opens the door to errors, such as forgetting to pop the stack after visiting a node, meaning that the stack could accidentally become out of sync with the actual syntax tree. Why do you prefer to build a new stack? I'm sure there's a good reason, so just asking to learn :)